support custom date format

shabegom opened the pull request:
This PR adds support for:
- custom date format for due and done dates
- toggle to make due and done dates links for linking to DNP

There are two new settings added to Tasks settings modal. One allows a date format to be set which is then used during moment.format(). The other is a toggle to make dates links.

I've also updated the due and done date regex to be a bit more open. I went the easy route instead of trying to guess all the different date formats. I hope this is ok!

TfTHacker commented:
+1 up vote for this feature

GitMurf commented:
+99 up vote as well ;-) This is great! Thanks @shabegom for doing this and thanks @schemar for the amazing plugin!

arminta7 commented:
+1

kmaustral commented:
+1

PJJPtx commented:
+1

schemar commented:
Thank you @shabegom :tada: 

I will do a proper review when I find the time. Until then, can you please:

1. Make sure tests and linting pass
2. Add documentation to the README
3. Consider whether there is a better way to set the format on the Task from the settings (maybe only once when the plugin initializes? Maybe get the format always from the settings? Setting the static property doesn't look right in my opinion)

Thank you for the effort! :heart:

schemar commented:
I think you can remove the `dateFormat` property from the Task class as you will get the format from the settings every time.

However, I can't wrap my head around a fundamental problem when allowing users to change the format of dates: what happens to all the tasks already in the vault? Their due and done dates would simply become invalid and instead be part of the description now. In your current implementation this also means the parser stops parsing metadata from the end of the task line and thus the recurrence rule would be lost as well.

In short: just allowing users to change the format will lead to what they deem a "broken" Tasks plugin and get back to me via bug tickets.

We need to consider updating all existing tasks when changing the setting. :thinking: *Or* we only allow a certain set of formats that we all try to match (regardless of the setting) when reading the task. Only newly written tasks would then have the new format, e.g. when toggling, editing, or creating a task.

What do you think?

shabegom commented:
Here's the simplest approach imo. I could add a button to the settings or a command that runs a one time conversion to any new date format specified. That way all previous tasks would get updated. it could also just auto update, but we'd need to validate the user has completed entering their new format into the field with onBlur or something...

I'm not too sure about accepting any date format, it's totally doable (just don't provide a second argument to moment() and any valid dateformat will be parsed). But I think it will confuse people if they see a different format between edit and preview of the tasks.

kmaustral commented:
That would be a big gesture. For me, I couldn't see any problem taking
personal responsibility for changing the date format for my previously
entered tasks, given it was my choice to set the option for a date link.

On Tue, 31 Aug 2021 at 05:23, Sam ***@***.***> wrote:

> Here's the simplest approach imo. I could add a button to the settings or
> a command that runs a one time conversion to any new date format specified.
> That way all previous tasks would get updated. it could also just auto
> update, but we'd need to validate the user has completed entering their new
> format into the field with onBlur or something...
>
> I'm not too sure about accepting any date format, it's totally doable
> (just don't provide a second argument to moment() and any valid dateformat
> will be parsed). But I think it will confuse people if they see a different
> format between edit and preview of the tasks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/schemar/obsidian-tasks/pull/296#issuecomment-908622961>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARTR6N2SOVROSJ73DH4C3XLT7PLEJANCNFSM5CXJTPAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jan-willi commented:
While going down the rabbit hole to modify this plugin to be my perfect task manager I stumbled across the same problem of being able to read multiple formats.
I "solved" it by allowing to set multiple formats in the settings. All formats can be read, while the first supplied format will be used for new tasks. Is this something that could work?

The branch I'm working on (not at all pretty enough for a PR):
https://github.com/jan-willi/obsidian-tasks/tree/datetime 
(Also includes custom signifiers and optional times in addition to dates. I might open separate PR's if this works out.)

TfTHacker commented:
Hoping to see this merge already for days :-)

I am also willing to migrate my tasks to the new format.

schemar commented:
I don't have time right now for private reasons. I plan to add some new things, including open PRs, to Tasks some time during October. Please be patient.

I think the best solution will be to not support *any* format, but a number of specific formats, including links (`[[]]`).

schemar commented:
Thank you for the update @shabegom. This is next on my list.

shabegom commented:
The fork is still a broken given the massive update. I'm working on it though. Totally didn't realize I opened this PR off the main branch...

schemar commented:
> The fork is still a broken given the massive update. I'm working on it though. Totally didn't realize I opened this PR off the main branch...

Thank you! If it is too much, I can also do it. I suspect I will do some changes after merging either way 👍

schemar commented:
It also looks like there are some unrelated changes @shabegom :thinking:

shabegom commented:
Do you have an example? I only changed things to get the custom date format and date links working with the new options. Anything else must have slipped in accidentally.

shabegom commented:
@schemar I'll revert those changes back today. Sorry about that. Just blindly following the linting errors.

schemar commented:
If the linter complains, it's fine to update 😄

shabegom commented:
@schemar I've reverted some of the unrelated changes and left the ones that were throwing errors. Also updated the tests in Query.test.ts.

I'm getting a test failure that I don't think is related to the changes:

Jest: concurrent test "Query sorting instructions sorting as {"input":"sort by status\nsort by due","output":[{"property":"status","reverse":false
},{"property":"due","reverse":false}]}" must return a Promise.

Let me know if there is anything else I can do to get this merge ready. Thanks!

schemar commented on src/QueryRenderer.ts:
```diff
@@ -118,11 +119,37 @@ class QueryRenderChild extends MarkdownRenderChild {
     }
 
     private async render({ tasks, state }: { tasks: Task[]; state: State }) {
+        tasks.map(async (task) => {
+            if (task.description.startsWith('![[')) {
+                const link = parseLinktext(task.description);
+                const subpath = link.subpath
+                    .replace('#^', '')
+                    .replace(']]', '');
+                if (link) {
+                    const file = this.app.metadataCache.getFirstLinkpathDest(
+                        link.path.replace('![[', ''),
+                        task.path,
+                    );
+                    const content =
+                        file && (await this.app.vault.read(file)).split('\n');
+                    const blocks =
+                        file &&
+                        this.app.metadataCache.getFileCache(file)?.blocks;
+                    const line = blocks && blocks[subpath]?.position.start.line;
+                    if (line) {
+                        task.description = content[line]
+                            .split(' ^')[0]
+                            .replace('- ', '');
+                    }
+                }
+            }
+        });
```
@shabegom can you please explain how this is related to the custom date format?

shabegom replied:
This is so weird! This is from a totally unrelated plugin (block ref counter) and I have no idea how it made it in here. My best guess is Github Copilot suggested this code and I hit tab by accident...

I'm sorry I didn't catch this before. That's really embarrassing. can be fully deleted.

schemar replied:
No worries

schemar commented on src/QueryRenderer.ts:
```diff
@@ -146,6 +172,7 @@ class QueryRenderChild extends MarkdownRenderChild {
     }: {
         tasks: Task[];
         content: HTMLDivElement;
+        app: App;
```
is this actually used?

shabegom replied:
I don't believe it is used. I missed this.

schemar reviewed (commented):
Thanks a bunch @shabegom! I will use parts of this to get me a speedy start with the feature. It really helps a lot. I will do some things differently. For example, I am not sure I cannot maintain arbitrary date formats.

schemar reviewed (approved):
Merging into a branch. It would still help me if you replied to my questions :smile:  Thank you!

schemar commented:
@shabegom, @TfTHacker, @GitMurf, @arminta7, @kmaustral, @PJJPtx, @jan-willi, @wenlzhang all of you upvoted or participated here or in the discussion #69.

If Tasks supported a fixed list of date formats (only dates, no times), which format(s) would you require? All would have the option to be links.

TfTHacker commented:
Thank you for your work on this project!!! Format for me:

YYYY-MM-DD

By the way, one of things I have done with my plugins  is get the date format for daily notes page which tends to be the format used by the user in his vault. This can be easily retrieved with this library. https://github.com/liamcain/obsidian-daily-notes-interface

schemar commented:
> By the way, one of things I have done with my plugins is get the date format for daily notes page which tends to be the format used by the user in his vault. This can be easily retrieved with this library. https://github.com/liamcain/obsidian-daily-notes-interface

This is great! Thank you!

kwjfef commented:
Please could you add the option to change the date format to DD-MM-YYYY (reverse order) to follow EU/UK regional formatting?

schemar commented:
> Please could you add the option to change the date format to DD-MM-YYYY (reverse order) to follow EU/UK regional formatting?

Will do 👍
